compiler/rustc_codegen_ssa/src/codegen_attrs.rs

```diff
@@ -608,11 +610,55 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, did: LocalDefId) -> CodegenFnAttrs {
     // strippable by the linker.
     //
     // Additionally weak lang items have predetermined symbol names.
-    if let Some((name, _)) = lang_items::extract(attrs)
+    if let Some((name, span)) = lang_items::extract(attrs)
         && let Some(lang_item) = LangItem::from_name(name)
     {
+        let has_no_mangle = codegen_fn_attrs.flags.contains(CodegenFnAttrFlags::NO_MANGLE);
         if WEAK_LANG_ITEMS.contains(&lang_item) {
             codegen_fn_attrs.flags |= CodegenFnAttrFlags::RUSTC_STD_INTERNAL_SYMBOL;
+            if has_no_mangle {
+                // This should never use the default as `no_mangle` is present in the attributes list
+                let no_mangle_span = no_mangle_span.unwrap_or_default();
+                let label_attr = format!("this defines it as the `{}` item", lang_item.name());
+                let label = format!("is the {} item", lang_item.name());
+                let mut err = tcx.dcx()
+                    .struct_span_err(no_mangle_span, "`#[no_mangle]` cannot be used on weak language items")
+                    .with_note("The linker requires specific names for internal language items,")
+                    .with_note("like panic handlers, eh personality functions or eh catch typeinfo variables.")
```

Comment:
This might be too specialized, but I imagine that many folks putting `no_mangle` on internal functions are doing so for debugging purposes. It might be nice to [include the debugging advice from the original issue](https://github.com/rust-lang/rust/issues/139923#issuecomment-2815568150) in the compiler hint here, i.e.:

```suggestion
                    .with_note("The linker requires specific names for internal language items,")
                    .with_note("like panic handlers, eh personality functions or eh catch typeinfo variables.")
                    .with_note("If you are trying to prevent mangling to ease debugging, many")
                    .with_note("debuggers support a command such as `rbreak {sym}` to")
                    .with_note("match `.*{sym}.*` instead of `break {sym}` on a specific name")
```

Reply:
Thanks @ppannuto.